Update instructors and requirements to 2024

lagru opened the pull request.

lagru committed:
Update year and links to 2024

lagru committed:
Update instructors

lagru committed:
Update requirements.txt

lagru commented on environment.yml:
I'm not sure what's the culprit, but when I setup this environment I get a segmenation fault when running the `test-env.py` script! I think it happens when it tries to use Napari. Is `napari-animation` the cause? I tried both the pip and conda-forge versions, same result...

psobolewskiPhD replied:
Any idea where it's segfaulting?
Does just `napari` run ?

lagru replied:
Yes. I narrowed it down to the line `labels_layer = viewer.add_labels(seg1)`. When I comment that one out everything runs fine. Trying to run it: `Segmentation fault`.

lagru replied:
> What OS are you using? I think we were getting different errors. Mine wasn't segfault.
It's working for me on OSX and Linux.

@JoOkuma, I'm on Arch Linux using micromamba. Maybe the latter is the reason? Trying with conda now instead..

lagru replied:
Same result with conda. 

<details><summary>napari --info</summary>

```
napari --info
napari: 0.4.19.post1
Platform: Linux-6.9.6-arch1-1-x86_64-with-glibc2.39
System: Arch Linux
Python: 3.12.4 | packaged by conda-forge | (main, Jun 17 2024, 10:23:07) [GCC 12.3.0]
Qt: 5.15.8
PyQt5: 5.15.9
NumPy: 1.26.4
SciPy: 1.14.0
Dask: 2024.6.2
VisPy: 0.14.3
magicgui: 0.8.3
superqt: 0.6.7
in-n-out: 0.2.1
app-model: 0.2.7
npe2: 0.7.6

OpenGL:
- GL version:  4.6 (Compatibility Profile) Mesa 24.1.2-arch1.1
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1707x960, scale 1.5

Settings path:
- /home/lg/.config/napari/image-analysis-24_c54568acd14b98a393c7245e4a51b51d186482c5/settings.yaml
Plugins:
- napari: 0.4.19.post1 (77 contributions)
- napari-animation: 0.0.8 (2 contributions)
- napari-console: 0.0.9 (0 contributions)
- napari-svg: 0.2.0 (2 contributions)
```

</details>

lagru replied:
Ha, weird. If I install everything via `pip install -r requirements.txt`, it works. The only difference I can see from `napari --info` is 

```patch
--- with conda
+++ with pip
@@ -1,8 +1,8 @@
-Qt: 5.15.8
-PyQt5: 5.15.9
+Qt: 5.15.2
+PyQt5: 5.15.10

```

psobolewskiPhD replied:
huh. So it's downgrading Qt, but upgrading PyQt? very odd.
Can you try 3.11? maybe it's related to python 3.12 migrations?

lagru replied:
Same behavior (segfault) with Python 3.11 and 3.10.

lagru replied:
But it works if I remove `cellpose`... :sweat_smile:  That's it, I'm removing that dependency for now and will add a comment to the exercise to install it after the fact if it's wanted.

lagru replied:
Note, that it works with the exact same output for `napari --info` as the failing one above. :shrug:

lagru replied:
When comparing the working env without cellpose and the segfaulting one with cellpose I can see that the following 2 deps are different (working vs failing):

- ffmpeg 7.0.1 vs 6.1.1
- pcre2 10.44 vs 10.43

lagru replied:
But downgrading both manually in the working environments doesn't produce a segfault.

psobolewskiPhD replied:
🤯 
I've definitely hit seg faults before with torch and import order...
But the test script doesn't even import it does it?
Our labels code does do some dtype normalization so maybe it's hitting some torch code....

psobolewskiPhD replied:
It's probably not even torch per se but some library that is used by napari and/or qt and/or torch.

lagru replied:
It doesn't import it. It's really weird, but hopefully not an issue for most tutorial participants now. :crossed_fingers: 

Thanks for debugging with me! :blush:

JoOkuma commented on environment.yml:
```diff
@@ -1,20 +1,20 @@
+  - seaborn=0.13.2
+  - napari=0.4.19.post1
   - cookiecutter
   - cellpose
```
```suggestion
  - cellpose
  - pyqt
```

psobolewskiPhD replied:
💯 
for conda install of napari, need to specify QT backend now, just like with pip ([all] includes pyqt)
I'd stick with pyqt, it's the more robust one.

JoOkuma commented:
Hey, @lagru, thanks for updating the older tutorial.

I suggested to include `pyqt` as a dependency.

Once I installed it`test-env.py` worked.

JoOkuma commented:
@lagru, What OS are you using? I think we were getting different errors. Mine wasn't segfault.
It's working for me on OSX and Linux.

psobolewskiPhD commented on tutorial/install_dependencies.md:
```diff
@@ -94,7 +94,7 @@ as well and you can skip to the next section.
    This ensures that the requirements for this workshop do not interfere with
    your other Python projects. To create the environment (named
-   `image-analysis-23`) and install Python 3.10 in it, enter the following command:
+   `image-analysis-24`) and install Python 3.10 in it, enter the following command:
```
```suggestion
   `image-analysis-24`) and install Python 3.12 in it, enter the following command:
```

psobolewskiPhD commented on tutorial/install_dependencies.md:
````diff
@@ -124,26 +124,26 @@ Sometimes, `napari` installation can fail on an M1 Mac due to mismatching
 
    ```bash
-   conda create -y -n image-analysis-23 -c conda-forge python=3.10 napari
+   conda create -y -n image-analysis-24 -c conda-forge python=3.10 napari
````
```suggestion
   conda create -y -n image-analysis-24 -c conda-forge python=3.12 napari pyqt
```
the yml uses 3.12 and as noted above pyqt is the best bet for a qt backend

psobolewskiPhD commented on environment.yml:
```diff
@@ -1,20 +1,20 @@
+  - napari=0.4.19.post1
   - cookiecutter
   - cellpose
   - pip:
```
Not sure why napari-animation is being installed from pip.
it's on conda-forge and it's safer to install that way because ffmpeg.

psobolewskiPhD replied:
sorry won't let me suggest the fix

lagru replied:
Not sure why we did it this way during the last tutorial. Maybe the conda-forge one was out of date? I'll update it to conda-forge. :+1:

psobolewskiPhD commented:
With adding pyqt and switching napari-animation to conda from pip, the test script runs on my mac -- I'm on M1 though.
BTW, I would suggest using arm64 miniforge these days on arm64 Macs instead of x86, but maybe there was a specific reason it was x86 before?

lagru committed:
Apply suggestions from code review

Co-authored-by: Peter Sobolewski <76622105+psobolewskiPhD@users.noreply.github.com>

lagru commented on requirements.txt:
```diff
@@ -2,13 +2,13 @@ jupyterlab
+napari==0.4.19.post1
 pyqt5
 cookiecutter
 cellpose
```
Considering that we probably won't be using the `04_deep_learning_segmentation.md` exercise, cellpose and its torch dependency seems very over-the-top. Let's remove it and add a comment to the top of the example in case someone really wants to try it? Or just remove the exercise altogether. What do you think? :thinking:

JoOkuma replied:
@lagru I had s similar problem with cellpose and torch, if cellpose was imported first it didn't crash. When imported after torch it would seg fault.
I fixed by always setting cellpose as the first import.

lagru replied:
That's horrible! :sweat_smile: I heard that pytorch envs can be messy/problematic but never really had much exposure...

lagru committed:
Install napari-animation via conda-forge

and explicitly use pyqt.

lagru committed:
Update forgotten reference to 3.12

lagru committed:
Update pages workflow

Copied and updated from
https://github.com/napari/napari-workshop-template/blob/9f43764ec83714805801884656c5dae72ffb1436/.github/workflows/pages.yml

lagru committed:
Remove cellpose as a default dependency

lagru committed:
Recommend arm64 on Mac with M1

Co-authored-by: Peter Sobolewski <76622105+psobolewskiPhD@users.noreply.github.com>

lagru commented:
Okay, I think this is ready to merge now. :D

PS: @psobolewskiPhD I invited you to this org as well.

lagru commented:
I'm merging this now since I won't be able to contribute much work tomorrow and over the weekend. I'm hoping to get the pages for this repo working so that we can point participants at nice HTML instead of markdown on GitHub.

Please feel welcome to continue to give feedback here if there are still thinks to address. And thanks for the reviews so far. :)

JoOkuma commented:
FYI @lagru @psobolewskiPhD, in case this happens during the tutorial.
A linux wizard (@ziw-liu) from work mentioned that the segfault might be from different versions of llvm being used by different packages, installing `numba` from the anaconda `numba` could fix it.

ziw-liu commented:
> FYI @lagru @psobolewskiPhD, in case this happens during the tutorial. A linux wizard (@ziw-liu) from work mentioned that the segfault might be from different versions of llvm being used by different packages, installing `numba` from the anaconda `numba` could fix it.

Just adding more details: The official Numba builds on PyPI and the Numba channel for conda (`-c numba`) statically link LLVM so they might avoid this issue. See https://github.com/conda-forge/llvmlite-feedstock/issues/84.
